repository/src/main/java/org/alfresco/repo/search/impl/AbstractCategoryServiceImpl.java

```diff
@@ -1,623 +1,677 @@
+        final Map<QName, PropertyDefinition> aspectProperties = aspectDefinition.getProperties();
+        final Optional<QName> catProperty = aspectProperties.entrySet().stream()
+                //for backwards compatibility I'm leaving the part where we get custom category aspects
+                .filter(ap -> ContentModel.ASPECT_GEN_CLASSIFIABLE.isMatch(aspectName) || isValidCategoryTypeProperty(aspectName, ap))
+                .map(Map.Entry::getKey)
+                .findFirst();
+
+        return catProperty.map(cp -> {
+            final String field = "@" + cp;
+            final SearchParameters sp = new SearchParameters();
+            sp.addStore(storeRef);
+            sp.setQuery(cp + ":*");
+            //we only care about faceted results and don't need query results so we can limit them to minimum
+            sp.setMaxItems(1);
+            sp.setSkipCount(0);
+            final SearchParameters.FieldFacet ff = new SearchParameters.FieldFacet(field);
+            ff.setLimitOrNull(count < 0 ? null : count);
+            sp.addFieldFacet(ff);
+            return sp;
+        })
+                .orElseThrow(() -> new IllegalStateException("Aspect does not have category property mirroring the aspect name"));
+    }
+
+    /**
+     * Checks whether given aspect property definition is valid category property
+
+     * @param aspectName Aspect name
+     * @param propertyDef Aspect property definition.
+     * @return is valid category property
+     */
+    private boolean isValidCategoryTypeProperty(QName aspectName, Map.Entry<QName, PropertyDefinition> propertyDef)
+    {
+        return propertyDef.getKey().getNamespaceURI().equals(aspectName.getNamespaceURI()) &&
+                propertyDef.getKey().getLocalName().equals(aspectName.getLocalName()) &&
+                DataTypeDefinition.CATEGORY.equals(propertyDef.getValue().getDataType().getName());
+    }
+
+    @Override
+    @Experimental
+    public Optional<NodeRef> getRootCategoryNodeRef(final StoreRef storeRef)
+    {
+        final NodeRef rootNode = nodeService.getRootNode(storeRef);
+        final ChildAssociationRef categoryRoot = nodeService.getChildAssocs(rootNode, Set.of(ContentModel.TYPE_CATEGORYROOT)).stream()
+                .findFirst()
+                .orElseThrow(() -> new CategoryServiceException(NODE_WITH_CATEGORY_ROOT_TYPE_NOT_FOUND));
+        final List<ChildAssociationRef> categoryRootAssocs = nodeService.getChildAssocs(categoryRoot.getChildRef());
+        if (CollectionUtils.isEmpty(categoryRootAssocs))
+        {
+            throw new CategoryServiceException(CATEGORY_ROOT_NODE_NOT_FOUND);
+        }
+        return categoryRootAssocs.stream()
+                .filter(ca -> ca.getQName().equals(ContentModel.ASPECT_GEN_CLASSIFIABLE))
+                .map(ChildAssociationRef::getChildRef)
+                .findFirst();
+    }
+
+    private <T> PagingResults<T> createPagingResult(final List<T> associations, final boolean hasMoreItems)
```

Comment:
This method may not be neede if we used `ListBackedPagingResults` class where it is called.

Reply:
Fixed